Support for MySQL SSL options (`--mysql-ssl-cert`, `--mysql-ssl-key`, `--mysql-ssl-ca`)

frugan-dev opened the pull request:
# Pull Request: Add MySQL SSL Options

## Description

This pull request adds support for MySQL SSL options (`--mysql-ssl-cert`, `--mysql-ssl-key`, `--mysql-ssl-ca`), that allow users to specify the paths to SSL certificate, key, and CA certificate files when connecting to MySQL databases that require secure transport (`--require_secure_transport=ON`).
This enhancement aims to address the issue where connections using insecure transport are prohibited due to MySQL server settings, e.g. when this error message appears:

> ERROR    3159 (HY000): Connections using insecure transport are prohibited while --require_secure_transport=ON.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

These changes have been tested locally by connecting to MySQL databases with `--require_secure_transport=ON` using the newly introduced SSL options.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

techouse commented:
Thanks for your contribution. 💪

Can you please fix the linter errors and add some extra tests for this feature?

frugan-dev commented:
Hi @techouse, ok I'll try in the next few days, be patient but I'm new to Python..

frugan-dev commented:
Hi @techouse, sorry for the delay, I fixed the linters errors and added new ssl tests.
Here are some considerations:

1) I added `tox.*.ini` in the .gitignore file, because on my pc the `mysql:latest` image did not run (see https://github.com/docker-library/mysql/issues/1055) and I had to create a `tox.override.ini` file:
```ini
[testenv]
allowlist_externals = pytest
commands = pytest -v --cov=src/mysql_to_sqlite3 --cov-report=xml --docker-mysql-image=mysql:8-oraclelinux8
```
and run the tests this way:
```sh
tox -c tox.ini -c tox.override.ini
```

2) Locally I had to use a value of `mysql_connection_retries: int = 50` [here](https://github.com/techouse/mysql-to-sqlite3/blob/master/tests/conftest.py#L199), as the default value was not sufficient; I have not committed the change, but perhaps it is a good idea to increase it, what do you think?

3) I captured the container logs [here](https://github.com/techouse/mysql-to-sqlite3/blob/2a1c560e5e0806a740e2fd430248484c8cc87e08/tests/conftest.py#L328), because passing self-signed certificates to the container kept giving me the error `Lost connection to MySQL server during query` and I didn't understand why..
In the end I understood that it was just a permissions problem, which for now I solved like [this](https://github.com/techouse/mysql-to-sqlite3/blob/2a1c560e5e0806a740e2fd430248484c8cc87e08/tests/conftest.py#L229), but maybe there is a more elegant solution (option `user` [here](https://docker-py.readthedocs.io/en/stable/containers.html)?).

4) I added a [random suffix](https://github.com/techouse/mysql-to-sqlite3/blob/2a1c560e5e0806a740e2fd430248484c8cc87e08/tests/conftest.py#L306) to the name of the created containers to avoid errors like `Conflict. The container name "/pytest_mysql_to_sqlite3" is already in use by container "xxx". You have to remove (or rename) that container to be able to reuse that name.")`.

Let me know if that works out!

techouse commented:
Hi @frugan-dev 

Looks like there are still some [fundamental errors](https://github.com/techouse/mysql-to-sqlite3/actions/runs/10342153625/job/28634166678?pr=77#step:9:229) in your code.

You don't have to use tox to test it locally, running something like
```bash
pytest -v -x
```
should do the trick.

As for the Tox setup, please do not commit / git-track these changes as they are specific to your setup and do not reflect a wider problem.

techouse commented:
> Locally I had to use a value of mysql_connection_retries: int = 50 [here](https://github.com/techouse/mysql-to-sqlite3/blob/master/tests/conftest.py#L199), as the default value was not sufficient; I have not committed the change, but perhaps it is a good idea to increase it, what do you think?

Hmm, this was never an issue before. Maybe it's SSL-specific. You can try to commit the change to `50` but that seems obscenely high to me.

frugan-dev commented:
> Looks like there are still some [fundamental errors](https://github.com/techouse/mysql-to-sqlite3/actions/runs/10342153625/job/28634166678?pr=77#step:9:229) in your code.

Thanks, I made some fixes.
Local testing with Docker works, but on GitHub there would be some adjustments to be made to support self-signed certificates.

> As for the Tox setup, please do not commit / git-track these changes as they are specific to your setup and do not reflect a wider problem.

Sure, I only committed the `.gitignore` change, so as to allow the use of a `tox.override.ini` file locally.

techouse commented:
> Local testing with Docker works, but on GitHub there would be some adjustments to be made to support self-signed certificates.

Commit whatever you think is necessary to make the CI tests pass. :)

frugan-dev commented:
> > Locally I had to use a value of mysql_connection_retries: int = 50 [here](https://github.com/techouse/mysql-to-sqlite3/blob/master/tests/conftest.py#L199), as the default value was not sufficient; I have not committed the change, but perhaps it is a good idea to increase it, what do you think?
> 
> Hmm, this was never an issue before. Maybe it's SSL-specific. You can try to commit the change to `50` but that seems obscenely high to me.

Actually on my PC it connects after about 20-25 attempts, but for now let's leave it like this, there's no problem.. I wouldn't want the increase to affect the execution time of the CI tests.

> > Local testing with Docker works, but on GitHub there would be some adjustments to be made to support self-signed certificates.
> 
> Commit whatever you think is necessary to make the CI tests pass. :)

Ok I'll take a look ASAP!

techouse commented:
Superseded by #128
